Data/IP/Addr.hs

```diff
@@ -258,85 +259,186 @@ showIPv6 ip@(IP6 (a1,a2,a3,a4))
-    split2 [] = []
-    split2 x  = take 2 x : split2 (drop 2 x)
-    toWord32 [a1,a2] = fromIntegral $ shift a1 16 + a2
-    toWord32 _       = error "toWord32"
+    !x1 = fromIntegral $ (i1 .&. 0xffff) `unsafeShiftL` 16 .|. (i2 .&. 0xffff)
+    !x2 = fromIntegral $ (i3 .&. 0xffff) `unsafeShiftL` 16 .|. (i4 .&. 0xffff)
+    !x3 = fromIntegral $ (i5 .&. 0xffff) `unsafeShiftL` 16 .|. (i6 .&. 0xffff)
+    !x4 = fromIntegral $ (i7 .&. 0xffff) `unsafeShiftL` 16 .|. (i8 .&. 0xffff)
+toIPv6 _ = error "toIPv6 field list length != 8"
+{-# INLINE toIPv6 #-}
 
 {-|
-  The 'toIPv6b' function takes a list of 'Int'
-  where each member repserents a single byte and returns 'IPv6'.
+  The 'toIPv6b' function returns the IPv6 address corresponding to the given
+  list of sixteen 'Int' octets.  The function is strict in the sixteen elements
+  of the list.  An error is returned if the list has a differnet length.  The
+  input elements are silently truncated to their 8 least-signicant bits before
+  they are combined to form the IPv6 address.
 
 >>> toIPv6b [0x20,0x01,0xD,0xB8,0,0,0,0,0,0,0,0,0,0,0,1]
 2001:db8::1
 -}
 toIPv6b :: [Int] -> IPv6
-toIPv6b ad = IP6 (x1,x2,x3,x4)
+toIPv6b [ h11, h12, l11, l12, h21, h22, l21, l22
+        , h31, h32, l31, l32, h41, h42, l41, l42 ] = IP6 (x1,x2,x3,x4)
   where
-    [x1,x2,x3,x4] = map toWord32 $ split4 ad
-    split4 [] = []
-    split4 x  = take 4 x : split4 (drop 4 x)
-    toWord32 [a1,a2,a3,a4] = fromIntegral $ shift a1 24 + shift a2 16 + shift a3 8 + a4
-    toWord32 _       = error "toWord32"
+    !x1 = fromIntegral $ (h11 .&. 0xff) `unsafeShiftL` 24 .|.
+                         (h12 .&. 0xff) `unsafeShiftL` 16 .|.
+                         (l11 .&. 0xff) `unsafeShiftL`  8 .|.
+                         (l12 .&. 0xff)
+    !x2 = fromIntegral $ (h21 .&. 0xff) `unsafeShiftL` 24 .|.
+                         (h22 .&. 0xff) `unsafeShiftL` 16 .|.
+                         (l21 .&. 0xff) `unsafeShiftL`  8 .|.
+                         (l22 .&. 0xff)
+    !x3 = fromIntegral $ (h31 .&. 0xff) `unsafeShiftL` 24 .|.
+                         (h32 .&. 0xff) `unsafeShiftL` 16 .|.
+                         (l31 .&. 0xff) `unsafeShiftL`  8 .|.
+                         (l32 .&. 0xff)
+    !x4 = fromIntegral $ (h41 .&. 0xff) `unsafeShiftL` 24 .|.
+                         (h42 .&. 0xff) `unsafeShiftL` 16 .|.
+                         (l41 .&. 0xff) `unsafeShiftL`  8 .|.
+                         (l42 .&. 0xff)
+toIPv6b _ = error "toIPv6b field list length != 16"
+
+{-|
+  The 'toIPv6w' function constructs the 'IPv6' address corresponding to the
+  given four-tuple of host byte order 'Word32' values.  This function differs
+  from the 'fromHostAddress6' function only in the fact that it is strict in
+  the elements of the tuple.
+
+>>> toIPv6w (0x20010DB8,0x0,0x0,0x1)
+2001:db8::1
+
+@since 1.7.9
+-}
+toIPv6w :: (Word32, Word32, Word32, Word32) -> IPv6
+toIPv6w w@(!_, !_, !_, !_) = IP6 w
```

Comment:
I think it is probably best to stay with the tuple, because that matches HostAddress6, and so conversion is free, no new heap objects need to be allocated. Also I can get the IPv6 address as a tuple again without memory allocation.

The "Data.IP.Internal" module exposes the `IP4(..)` and `IP6(..)` constructors, so changing it would also be a compatibility break.